flink-core/src/main/java/org/apache/flink/api/common/accumulators/ContinuousHistogram.java

```diff
@@ -0,0 +1,534 @@
+
+	protected TreeMap<Double, Integer> treeMap = new TreeMap<Double, Integer>();
+
+	protected long counter = 0;
+
+	private int bin;
+
+	private double lower;
+
+	private double upper;
+
+	private PriorityQueue<KeyDiff> diffQueue;
+
+	private HashMap<Double, KeyProps> keyUpdateTimes;
+
+	private long timestamp;
+
+	/**
+	 * Creates a new Continuous histogram with the given number of bins
+	 * Bins represents the number of values the histogram stores to approximate the continuous
+	 * data set. The higher this value, the more we move towards an exact representation of data.
+	 *
+	 * @param numBins Number of bins in the histogram
+	 */
+	public ContinuousHistogram(int numBins) {
+		if (numBins <= 0) {
+			throw new IllegalArgumentException("Number of bins must be greater than zero");
+		}
+		bin = numBins;
+		lower = MAX_VALUE;
+		upper = -MAX_VALUE;
+		diffQueue = new PriorityQueue<>();
+		keyUpdateTimes = new HashMap<>();
+		timestamp = 0;
+	}
+
+	/**
+	 * Consider using {@link #add(double)} for primitive double values to get better performance.
+	 */
+	@Override
+	public void add(Double value) {
+		add(value, 1);
+	}
+
+	public void add(double value) {
+		add(value, 1);
+	}
+
+	@Override
+	public TreeMap<Double, Integer> getLocalValue() {
+		return this.treeMap;
+	}
+
+	/**
+	 * Get the total number of items added to this histogram.
+	 * This is preserved across merge operations.
+	 *
+	 * @return Total number of items added to the histogram
+	 */
+	public long getTotal() {
+		return counter;
+	}
+
+	/**
+	 * Get the current size of the {@link #treeMap}
+	 *
+	 * @return Size of the {@link #treeMap}
+	 */
+	public int getSize() {
+		return treeMap.size();
+	}
+
+	@Override
+	public void resetLocal() {
+		treeMap.clear();
+		counter = 0;
+		lower = MAX_VALUE;
+		upper = -MAX_VALUE;
+		diffQueue.clear();
+		keyUpdateTimes.clear();
+	}
+
+	@Override
+	public void merge(Accumulator<Double, TreeMap<Double, Integer>> other) {
+		fill(other.getLocalValue().entrySet());
+	}
+
+	/**
+	 * Merges the given other histogram into this histogram, with the number of bins in the
+	 * merged histogram being {@code numBins}.
+	 *
+	 * @param other   Histogram to be merged
+	 * @param numBins Bins in the merged histogram
+	 */
+	public void merge(Accumulator<Double, TreeMap<Double, Integer>> other, int numBins) {
+		bin = numBins;
+		merge(other);
+	}
+
+	@Override
+	public Accumulator<Double, TreeMap<Double, Integer>> clone() {
+		ContinuousHistogram result = new ContinuousHistogram(bin);
+		result.treeMap = new TreeMap<>(treeMap);
+		result.counter = counter;
+		result.lower = lower;
+		result.upper = upper;
+		// initialize all differences and key update times for the new histogram
+		result.computeDiffs();
+		return result;
+	}
+
+	void add(double value, int count) {
+		addValue(value, count);
+		if (getSize() > bin) {
+			mergeBins();
+		}
+	}
+
+	void fill(Set<Map.Entry<Double, Integer>> entries) {
```

Comment:
This is for testing purposes. I use the `fill` method to initialize the histogram directly from a Map of values in test classes. Besides this will remain package private.